index.js

```diff
@@ -1,174 +1,175 @@
-      if (!this.opts.cid) return true
-      const cid = await this.cid()
-      const data = this.encodeUnsafe()
-      // https://github.com/bcoe/c8/issues/135
-      /* c8 ignore next */
-      return multihash.validate(cid.multihash, data)
+  decode () {
+    const decoded = this.decodeUnsafe()
+    if (decoded === null) return null
+    if (isBinary(decoded)) return copyBinary(decoded)
+    if (immutableTypes.has(typeof decoded) || decoded === null) {
+      return decoded
     }
+    return clone(decoded)
+  }
 
-    _encode () {
-      this._encoded = this.opts.data || multicodec.get(this.code).encode(this.opts.source)
-    }
+  encodeUnsafe () {
+    if (this._data) return this._data
+    throw new Error('Block created without an encoded state')
+  }
 
-    encode () {
-      if (!this._encoded) this._encode()
-      return copyBinary(this._encoded)
-    }
+  encode () {
+    const data = this.encodeUnsafe()
+    return copyBinary(data)
+  }
 
-    encodeUnsafe () {
-      if (!this._encoded) this._encode()
-      return this._encoded
-    }
+  async cid () {
+    if (this._cid) return this._cid
+    const hash = await this._hasher.digest(this.encodeUnsafe())
+    const cid = CID.create(1, this.opts.codec.code, hash)
+    setImmutable(this, '_cid', cid)
+    return cid
+  }
 
-    _decode () {
-      if (typeof this.opts.source !== 'undefined') this._decoded = this.opts.source
-      else {
-        const { decode } = multicodec.get(this.code)
-        this._decoded = decode(this._encoded || this.opts.data)
-      }
-      return this._decoded
-    }
+  get code () {
+    if (this._cid) return this._cid.code
+    return this._codec.code
+  }
 
-    decode () {
-      // TODO: once we upgrade to the latest data model version of
-      // dag-pb that @gozala wrote we should be able to remove this
-      // and treat it like every other codec.
-      /* c8 ignore next */
-      if (this.codec === 'dag-pb') return this._decode()
-      if (!this._decoded) this._decode()
-      const tt = typeof this._decoded
-      if (tt === 'number' || tt === 'boolean') {
-        // return any immutable types
-        return this._decoded
-      }
-      if (isBinary(this._decoded)) return copyBinary(this._decoded)
-      if (immutableTypes.has(typeof this._decoded) || this._decoded === null) {
-        return this._decoded
-      }
-      return clone(this._decoded)
-    }
+  reader () {
+    return reader(this.decodeUnsafe())
+  }
 
-    decodeUnsafe () {
-      if (!this._decoded) this._decode()
-      return this._decoded
-    }
+  async equals (block) {
+    if (block === this) return true
+    if (block.asBlock !== block) return false
+    const [a, b] = await Promise.all([this.cid(), block.cid()])
+    return a.equals(b)
+  }
+}
 
-    reader () {
-      return reader(this.decodeUnsafe())
+class BlockDecoder extends Block {
+  decodeUnsafe () {
+    if (typeof this.opts._source !== 'undefined') return this._source
+    if (!this._codec) {
```

Comment:
This check is obsolete because validation occurs at construction site.

Reply:
turns out, there is a case in which you want to create a Block without a codec attached. i actually have to drop the constructor error and move that check to the encoder/decoder methods, but it’ll still be necessary here so that you get a good error if you try to decode a block that doesn’t have a codec attached